### ISSUE TITLE
Regression: submodule access only generates an error with warm cache

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#20174]
This incremental mode test case (added to `check-incremental.test`) started generating a false positive after e1643aec5229e37cf62700dabae294b9a5344c97:
```py
[case testIncrementalAccessSubmoduleWithoutExplicitImport]
import b
import a

[file a.py]
import pkg

pkg.submod.foo()

[file a.py.2]
import pkg

pkg.submod.foo()
x = 1

[file b.py]
import c

[file c.py]
from pkg import submod

[file pkg/__init__.pyi]
[file pkg/submod.pyi]
def foo() -> None: pass
```
It fails like this:
```
---------------------------------------------------- Captured stderr call ----------------------------------------------------
Expected:
Actual:
  tmp/a.py:3: error: "object" has no attribute "submod" (diff)
```
On the previous commit (9c26271945ee7c9095d769f2138f4f2e969445ce) the test passes.

This also generated false positives in an internal codebase at work when we tried using a recent mypy development version.

The original behavior is questionable, but different semantics in incremental and non-incremental modes is a much bigger issue.

Credits to @p-sawicki for finding the commit that introduced this.

cc @ilevkivskyi

**Comment by ilevkivskyi** [comment]
Hm, maybe the original behavior is also causing https://github.com/python/mypy/issues/20079 (which is a problem for `scipy`)? If this is the case, then I would argue we should fix the original behaviour instead (so that cold runs behave like warm runs, not vice versa).

**Comment by JukkaL** [comment]
Yeah, the new warm run behavior seems more consistent with expectations. However, the new behavior might generate many new errors, in which case we might want to introduce it behind a feature flag if at all possible.

**Comment by ilevkivskyi** [comment]
That said, the fix for this issue may be much simpler (order the dependencies by the import line on warm runs, to match the original order). I will play with this a bit.

**Comment by ilevkivskyi** [comment]
It is actually worse than that. It looks like right now it works by pure chance. After
```
[file a.py]
import pkg
pkg.submod
```
we only record that `a` depends on `pkg`, and that's it, `a` _does not_ have `pkg.submod` as a dependency. With old logic `pkg.submod` was loaded because we used to load all fresh SCCs encountered so far, not just the dependencies of a stale SCC we are currently processing.

I am not sure what to do here. Making `pkg.submod` a dependency of `a` will break the idea that we can determine dependencies using only "first pass" semantic analysis. An alternative would be to always make `pkg.submod` a (low priority) dependency of `pkg` (not sure how easy it is).

**Comment by JukkaL** [comment]
The old behavior seems to have worked consistently, even if it seems odd -- I don't recall any issues where a submodule couldn't be resolved only in incremental mode. I don't know how it worked exactly to make this deterministic, however. 
How hard would it be to replicate the old logic in the new implementation -- or would it have some major drawbacks?

Another way would be to only allow resolving submodules that are imported by any transitive dependencies. So we wouldn't make `pkg` always depend on `pkg.submod`, but only if it imports `pkg.submod`. Similarly if `a` imports `b` which imports `pkg.submod`, this would make `pkg.submod` accessible in `a`. I think this would correspond to how things work at runtime. In `a`, a reference to `pkg.submod` would resolve if somewhere in transitive dependencies there is an import of `pkg.submod` (possibly in `pkg` but it could by somewhere else as well).

A stricter approach would be to require all used submodules to be imported explicitly in the module that uses them, unless they are explicitly exported by the outer package (e.g. `pkg` having `from . import submod`), but this may cause a significant backward compatibility break, and we'd need to experiment with this first.

**Comment by ilevkivskyi** [comment]
> How hard would it be to replicate the old logic in the new implementation -- or would it have some major drawbacks?

To re-iterate: the old logic was to load _all_ fresh modules that appear before current SCC in global topological sort order (this makes `visit_member_expr()`, or more precisely `get_module_symbol()`, work even though the symbol for `pkg.submod` is _not_ in the `pkg` symbol table). First, this is a waste of resources. It may be not so bad now, but will be worse with parallel checking. Second, this feels conceptually wrong, as it may hide more bugs. We should not rely on anything except dependencies being loaded at any point.

> Another way would be to only allow resolving submodules that are imported by any transitive dependencies

To be clear, this will make the original repro fail on the cold run. Since `a` only imports `pkg`, and `pkg` doesn't import `pkg.resources`. Are we fine with making this (also technically breaking) change?

**Comment by ilevkivskyi** [comment]
> Are we fine with making this (also technically breaking) change?

To be clear, I am asking because this change is not too hard to implement  (unless I am missing something). Also I think the last option you propose is way too strict.

**Comment by JukkaL** [comment]
> To be clear, this will make the original repro fail on the cold run.

Yes. But if the impact isn't massive, I'd argue this to be a bug fix, since `pkg.submod` being available since some module in the build happens to import it seems like unexpected behavior. If the impact is massive, our only option may be to first implement the new logic behind a feature flag, and flip the default in a follow-up feature release.

**Comment by JukkaL** [comment]
> Also I think the last option you propose is way too strict.

You are probably right, so the middle option seems like the most reasonable way forward.

**Comment by BobTheBuidler** [comment]
This appears potentially related to (and the fix a potential fix for) https://github.com/mypyc/mypyc/issues/1155
